server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java

```diff
@@ -90,73 +93,83 @@
+        Optional<ImmutableSet<MessageHeaderProperty>> headerProperties = getMessagesRequest.getHeaderProperties();
         return Stream.of(JmapResponse.builder().clientId(clientId)
                             .response(getMessagesResponse(mailboxSession, getMessagesRequest, requestedProperties))
                             .responseName(RESPONSE_NAME)
-                            .properties(requestedProperties.map(this::handleSpecificProperties))
-                            .filterProvider(Optional.of(buildFilteringHeadersFilterProvider(requestedProperties)))
+                            .properties(handleSpecificProperties(requestedProperties, headerProperties))
+                            .filterProvider(Optional.of(buildFilteringHeadersFilterProvider(headerProperties)))
                             .build());
     }
 
-    private Set<MessageProperty> handleSpecificProperties(Set<MessageProperty> input) {
+    private Optional<Set<? extends Property>> handleSpecificProperties(Optional<ImmutableSet<MessageProperty>> requestedProperties, Optional<ImmutableSet<MessageHeaderProperty>> headerProperties) {
         Set<MessageProperty> toAdd = Sets.newHashSet();
         Set<MessageProperty> toRemove = Sets.newHashSet();
-        ensureContainsId(input, toAdd);
-        handleBody(input, toAdd, toRemove);
-        handleHeadersProperties(input, toAdd, toRemove);
-        return Sets.union(Sets.difference(input, toRemove), toAdd).immutableCopy();
+        toAdd.addAll(ensureContainsMandatoryFields(requestedProperties));
+        handleBody(requestedProperties, toAdd, toRemove);
+        handleHeadersProperties(headerProperties, toAdd, toRemove);
+        ImmutableSet<MessageProperty> resultProperties = Sets.union(
+                    Sets.difference(requestedProperties.isPresent() ? requestedProperties.get() : ImmutableSet.of(), toRemove)
+                    , toAdd)
+                .immutableCopy();
+        if (resultProperties.isEmpty()) {
+            return Optional.empty();
+        }
+        return Optional.of(resultProperties);
     }
-        
-    private void ensureContainsId(Set<MessageProperty> input, Set<MessageProperty> toAdd) {
-        if (!input.contains(MessageProperty.id)) {
-            toAdd.add(MessageProperty.id);
+
+    private void handleHeadersProperties(Optional<ImmutableSet<MessageHeaderProperty>> headerProperties, Set<MessageProperty> toAdd, Set<MessageProperty> toRemove) {
+        if (headerProperties.isPresent() && !headerProperties.get().isEmpty()) {
+            toAdd.add(MessageProperty.headers);
         }
     }
-    
-    private void handleBody(Set<MessageProperty> input, Set<MessageProperty> toAdd, Set<MessageProperty> toRemove) {
-        if (input.contains(MessageProperty.body)) {
+
+    private Set<MessageProperty> ensureContainsMandatoryFields(Optional<ImmutableSet<MessageProperty>> requestedProperties) {
```

Comment:
all this is overly complex, why don't you just write `Sets.difference(mandatory, request)`